AMM/contracts/AMM/tests/functions/add_pool.rs

```diff
@@ -0,0 +1,89 @@
+use crate::utils::{
+    amm_abi_calls::{add_pool, pool},
+    test_helpers::{deploy_and_construct_exchange_contract, setup, setup_and_initialize},
+};
+
+mod success {
+    use super::*;
+
+    #[tokio::test]
+    async fn adds() {
+        let (wallet, amm_instance, assets) = setup_and_initialize().await;
+        let pair = (assets[0], assets[1]);
+        let exchange_contract_id =
+            deploy_and_construct_exchange_contract(&wallet, pair, None).await;
+        add_pool(&amm_instance, pair, exchange_contract_id).await;
+
+        let exchange_contract_id_in_storage = pool(&amm_instance, pair).await;
+        assert_ne!(exchange_contract_id_in_storage, None);
+        assert_eq!(
+            exchange_contract_id_in_storage.unwrap(),
+            exchange_contract_id
+        );
+    }
+
+    #[tokio::test]
+    async fn adds_more_than_once() {
```

Comment:
I would reorganize the workflow to define at the top and test at the bottom so that the reader does not have to read the test line-by-line to find where something is defined. E.g. move `pair_2` below `pair_1` then `exchange...id_1` below that and `exchange..._id_2` below that. Then add the pools and then assert at the end.